File: CS/TreeViewCheckBoxVisibility/Helpers/MyHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using DevExpress.Web.Mvc;
using DevExpress.Web.ASPxTreeView;
using System.Collections;

namespace TreeViewCheckBoxVisibility.Helpers {
    public static class MyHelper {
        private static void PerformActionOnNodesRecursive(TreeViewNodeCollection nodes, Action<TreeViewNode> action) {
            foreach (TreeViewNode node in nodes) {
                action(node);
                if (node.Nodes.Count > 0)
                    PerformActionOnNodesRecursive(node.Nodes, action);
            }
        }
        public static void AllowCheckNodes(MVCxTreeView treeView, int selectedIndex) {
            switch (selectedIndex) {
                case 0:
                    treeView.AllowCheckNodes = false;
                    break;
                case 1:
                    treeView.AllowCheckNodes = true;
                    PerformActionOnNodesRecursive(treeView.Nodes, delegate(TreeViewNode node) { node.AllowCheck = node.Nodes.Count == 0; });
                    break;
                case 2:
                    treeView.AllowCheckNodes = true;
                    PerformActionOnNodesRecursive(treeView.Nodes, delegate(TreeViewNode node) { node.AllowCheck = true; });
                    break;
            }
        }
        public static string ChangeAccessibility(MVCxTreeView treeView, bool enable, string disabledList) {
            if (!enable) {
                ArrayList lst = new ArrayList();
                PerformActionOnNodesRecursive(treeView.Nodes, delegate(TreeViewNode node) { 
                    node.Enabled = !node.Checked;
                    if (!node.Enabled)
                       lst.Add(node.Name);
                });
                disabledList = string.Join(",", lst.ToArray());
            }
            else {
                PerformActionOnNodesRecursive(treeView.Nodes, delegate(TreeViewNode node) { node.Enabled = true; });
                disabledList = string.Empty;
            }
            return disabledList;
        }
        public static void RestoreSelection(MVCxTreeView treeView, string selectedNodes) {
            if (string.IsNullOrEmpty(selectedNodes))
                return;
            string[] strNodes = selectedNodes.Split(',');
            for (int i = 0; i < strNodes.Length; i++) {
                TreeViewNode node = treeView.Nodes.FindRecursive(n => n.Name == strNodes[i]);
                if (node != null)
                    node.Checked = true;
            }
        }
        public static void RestoreAccessibility(MVCxTreeView treeView, string disabledList) {
            if (string.IsNullOrEmpty(disabledList))
                return;
            string[] strNodes = disabledList.Split(',');
            for (int i = 0; i < strNodes.Length; i++) {
                TreeViewNode node = treeView.Nodes.FindRecursive(n => n.Name == strNodes[i]);
                if (node != null)
                    node.Enabled = false;
            }
        }
    }
}

File: VB/TreeViewCheckBoxVisibility_vb/Helpers/MyHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports DevExpress.Web.Mvc
Imports DevExpress.Web.ASPxTreeView
Imports System.Collections

Namespace TreeViewCheckBoxVisibility.Helpers
	Public Class MyHelper
		Private Sub New()
		End Sub
        Private Shared Sub PerformActionOnNodesRecursive(ByVal nodes As TreeViewNodeCollection, ByVal action As Action(Of TreeViewNode))
            For Each node As TreeViewNode In nodes
                action(node)
                If node.Nodes.Count > 0 Then
                    PerformActionOnNodesRecursive(node.Nodes, action)
                End If
            Next node
        End Sub
		Public Shared Sub AllowCheckNodes(ByVal treeView As MVCxTreeView, ByVal selectedIndex As Integer)
			Select Case selectedIndex
				Case 0
					treeView.AllowCheckNodes = False
				Case 1
					treeView.AllowCheckNodes = True
                    PerformActionOnNodesRecursive(treeView.Nodes, Sub(node)
                                                                      node.AllowCheck = (node.Nodes.Count = 0)
                                                                  End Sub)
				Case 2
					treeView.AllowCheckNodes = True
                    PerformActionOnNodesRecursive(treeView.Nodes, Sub(node)
                                                                      node.AllowCheck = True
                                                                  End Sub)
			End Select
		End Sub
		Public Shared Function ChangeAccessibility(ByVal treeView As MVCxTreeView, ByVal enable As Boolean, ByVal disabledList As String) As String
			If (Not enable) Then
				Dim lst As ArrayList = New ArrayList()
                PerformActionOnNodesRecursive(treeView.Nodes, Sub(node)
                                                                  node.Enabled = Not node.Checked
                                                                  If (Not node.Enabled) Then
                                                                      lst.Add(node.Name)
                                                                  End If
                                                              End Sub)
                disabledList = String.Join(",", lst.ToArray())
            Else
                PerformActionOnNodesRecursive(treeView.Nodes, Sub(node)
                                                                  node.Enabled = True
                                                              End Sub)
                disabledList = String.Empty
            End If
                Return disabledList
        End Function
		Public Shared Sub RestoreSelection(ByVal treeView As MVCxTreeView, ByVal selectedNodes As String)
			If String.IsNullOrEmpty(selectedNodes) Then
				Return
			End If
			Dim strNodes As String() = selectedNodes.Split(","c)
			For i As Integer = 0 To strNodes.Length - 1
                Dim node As TreeViewNode = treeView.Nodes.FindRecursive(Function(n) n.Name = strNodes(i))
				If Not node Is Nothing Then
					node.Checked = True
				End If
			Next i
		End Sub
		Public Shared Sub RestoreAccessibility(ByVal treeView As MVCxTreeView, ByVal disabledList As String)
			If String.IsNullOrEmpty(disabledList) Then
				Return
			End If
			Dim strNodes As String() = disabledList.Split(","c)
			For i As Integer = 0 To strNodes.Length - 1
                Dim node As TreeViewNode = treeView.Nodes.FindRecursive(Function(n) n.Name = strNodes(i))
				If Not node Is Nothing Then
					node.Enabled = False
				End If
			Next i
		End Sub
	End Class
End Namespace